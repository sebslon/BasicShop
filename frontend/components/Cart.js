import styled from "styled-components";

import { useUser } from "./User";

import formatMoney from "../lib/formatMoney";
import calcTotalCartPrice from "../lib/calcTotalCartPrice";

import { CartContainer, CartHeader } from "./styles/Cart";

export default function Cart() {
  const me = useUser();

  if (!me) return null;

  return (
    <CartContainer open>
      <header>
        <CartHeader>{me.name}'s cart</CartHeader>
      </header>

      <ul>
        {me.cart.map((cartItem) => (
          <CartItem key={cartItem.id} cartItem={cartItem} />
        ))}
      </ul>

      <footer>
        <p>{formatMoney(calcTotalCartPrice(me.cart))}</p>
      </footer>
    </CartContainer>
  );
}

function CartItem({ cartItem }) {
  const { product } = cartItem;

  if (!product) return null;

  return (
    <Item>
      <img
        width="100"
        src={product.photo.image.publicUrlTransformed}
        alt={product.name}
      />
      <div>
        <h3>{product.name}</h3>
        <p>
          {formatMoney(product.price * cartItem.quantity)} -
          <em>
            {" " + cartItem.quantity} &times; {formatMoney(product.price)} each
          </em>
        </p>
      </div>
    </Item>
  );
}

const Item = styled.li`
  padding: 1rem 0;
  border-bottom: 1px solid var(--secondary);
  display: grid;
  grid-template-columns: auto 1fr auto;

  img {
    margin-right: 1rem;
  }

  h3,
  p {
    margin: 0;
  }
`;
