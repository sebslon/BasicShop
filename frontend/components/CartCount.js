import styled from "styled-components";
import { CSSTransition, TransitionGroup } from "react-transition-group";

export default function CartCount({ count }) {
  return (
    <AnimationStyles>
      <TransitionGroup>
        <CSSTransition
          unmountOnExit
          className="count"
          classNames="count"
          key={count}
          timeout={{ enter: 300, exit: 300 }}
        >
          <Dot>{count}</Dot>
        </CSSTransition>
      </TransitionGroup>
    </AnimationStyles>
  );
}

const Dot = styled.div`
  background: var(--secondary);
  color: white;
  border-radius: 50%;
  padding: 0.5rem;
  line-height: 2rem;
  min-width: 3rem;
  margin-left: 1rem;

  font-feature-settings: "tnum";
  font-variant-numeric: tabular-nums;
`;

const AnimationStyles = styled.span`
  position: relative;

  .count {
    display: block;
    position: relative;
    transition: transform 0.3s;
    backface-visibility: hidden;
  }

  .count-enter {
    transform: scale(4) rotateX(0.5turn);
  }

  .count-enter-active {
    transform: rotateX(0);
  }

  .count-exit {
    top: 0;
    position: absolute;
    transform: rotateX(0);
  }

  .count-exit-active {
    transform: scale(4) rotateX(0.5turn);
  }
`;
