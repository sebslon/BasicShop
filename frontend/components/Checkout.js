import styled from "styled-components";
import nProgress from "nprogress";
import { useState } from "react";
import { loadStripe } from "@stripe/stripe-js";
import {
  CardElement,
  Elements,
  useElements,
  useStripe,
} from "@stripe/react-stripe-js";

import SubmitPaymentBtn from "./styles/SubmitPaymentBtn";

const CheckoutForm = styled.form`
  box-shadow: 0 1px 2px 2px rgba(0, 0, 0, 0.5);
  border: 1px solid rgba(0, 0, 0, 0.7);
  border-radius: 5px;
  padding: 1rem;
  display: grid;
  grid-gap: 1rem;
`;

const stripeLib = loadStripe(process.env.NEXT_PUBLIC_STRIPE_KEY);

function Checkout() {
  const [error, setError] = useState();
  const [loading, setLoading] = useState(false);
  const stripe = useStripe();
  const elements = useElements();

  async function handleSubmit(e) {
    e.preventDefault();
    setLoading(true);
    nProgress.start();

    const { error, paymentMethod } = await stripe.createPaymentMethod({
      type: "card",
      card: elements.getElement(CardElement),
    });

    if(error) setError(error);

    setLoading(false);
    nProgress.done();
  }

  return (
    <CheckoutForm onSubmit={handleSubmit}>
      {error && <p style={{ fontSize: 12 }}>{error.message}</p>}
      <CardElement />
      <SubmitPaymentBtn>Check Out Now</SubmitPaymentBtn>
    </CheckoutForm>
  );
}

export default function CheckoutWrapper() {
  return (
    <Elements stripe={stripeLib}>
      <Checkout></Checkout>
    </Elements>
  );
}
