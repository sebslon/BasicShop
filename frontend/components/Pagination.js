import { useQuery } from "@apollo/client";
import gql from "graphql-tag";
import Head from "next/head";
import Link from "next/link";

import { perPage } from "../config";
import DisplayError from "./ErrorMessage";
import PaginationContainer from "./styles/PaginationContainer";

const PAGINATION_QUERY = gql`
  query PAGINATION_QUERY {
    _allProductsMeta {
      count
    }
  }
`;

export default function Pagination({ page }) {
  const { error, loading, data } = useQuery(PAGINATION_QUERY);

  if (loading) return "Loading...";
  if (error) return <DisplayError error={error} />;

  const { count } = data._allProductsMeta;
  const pageCount = Math.ceil(count / perPage);

  return (
    <PaginationContainer>
      <Head>
        <title>
          BasicShop - Page {page} of {pageCount}
        </title>
      </Head>
      
      <Link href={`/products/${page - 1}`}>
        <a aria-disabled={page === 1}>← Prev</a>
      </Link>
      <p>
        Page {page} of {pageCount}
      </p>
      <p>Total items {count}</p>
      <Link href={`/products/${page + 1}`}>
        <a aria-disabled={page >= pageCount}>→ Next</a>
      </Link>
    </PaginationContainer>
  );
}
