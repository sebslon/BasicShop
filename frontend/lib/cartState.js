import { useState, createContext, useContext } from "react";

const LocalStateContext = createContext();
const LocalStateProvider = LocalStateContext.Provider;

function CartStateProvider({ children }) {
  const [cartOpen, setCartOpen] = useState(false);

  function toggleCart() {
    setCartOpen((cart) => !cart);
  }

  function closeCart() {
    setCartOpen(false);
  }

  function openCart() {
    setCartOpen(true);
  }

  return (
    <LocalStateProvider
      value={{ cartOpen, setCartOpen, toggleCart, closeCart, openCart }}
    >
      {children}
    </LocalStateProvider>
  );
}

function useCart() {
  const cart = useContext(LocalStateContext);

  return cart;
}

export { CartStateProvider, useCart };
