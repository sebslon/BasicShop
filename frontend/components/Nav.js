import Link from "next/link";

import { useUser } from "./User";
import { useCart } from "../lib/cartState";
import StyledNav from "./styles/StyledNav";

import CartCount from "./CartCount";
import SignOut from "./SignOut";

export default function Nav() {
  const user = useUser();
  const { openCart } = useCart();

  return (
    <StyledNav>
      <Link href="./products">Products</Link>
      {user && (
        <>
          <Link href="./sell">Sell</Link>
          <Link href="./orders">Orders</Link>
          <Link href="./account">Account</Link>
          <button type="button" onClick={openCart}>
            My Cart
            <CartCount
              count={user.cart.reduce(
                (total, cartItem) =>
                  total + (cartItem.product ? cartItem.quantity : 0),
                0
              )}
            />
          </button>
          <SignOut />
        </>
      )}

      {!user && (
        <>
          <Link href="/signin">Sign In</Link>
        </>
      )}
    </StyledNav>
  );
}
