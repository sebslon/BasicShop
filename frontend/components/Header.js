import Link from "next/link";

import Logo from './styles/Logo';
import StyledHeader from './styles/StyledHeader';

import Cart from './Cart';
import Nav from "./Nav";
import Search from "./Search";



export default function Header() {
  return (
    <StyledHeader>
      <div className="bar">
        <Logo>
          <Link href="/">Basic Shop</Link>
        </Logo>
        <Nav />
      </div>

      <div className="sub-bar">
        <Search />
      </div>

      <Cart />
    </StyledHeader>
  );
}
