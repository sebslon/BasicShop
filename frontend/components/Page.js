import PropTypes from "prop-types";
import styled from "styled-components";

import { GlobalStyles } from "../GlobalStyles";
import Header from "./Header";

const InnerStyles = styled.div`
  max-width: var(--maxWidth);
  margin: 0 auto;
  padding: 2rem;
`;

export default function Page({ children }) {
  return (
    <div>
      <GlobalStyles />
      <Header />
      <InnerStyles>{children}</InnerStyles>
    </div>
  );
}