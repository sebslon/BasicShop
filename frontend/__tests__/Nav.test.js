import { render, screen } from "@testing-library/react";
import { MockedProvider } from "@apollo/react-testing";

import Nav from "../components/Nav";
import { CURRENT_USER_QUERY } from "../components/User";
import { fakeUser, fakeCartItem } from "../lib/testUtils";
import { CartStateProvider } from "../lib/cartState";

//Mocks for being logged in/out/ logged in with cart items

const noSignedInMocks = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: { authenticatedItem: null },
  },
];

const signedInMocks = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: { data: { authenticatedItem: fakeUser() } },
  },
];

const signedInMocksWithCartItems = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: {
      data: {
        authenticatedItem: fakeUser({
          cart: [fakeCartItem()],
        }),
      },
    },
  },
];

describe("<Nav />", () => {
  it("Renders minimal nav and sign in link when signed out", () => {
    const { container } = render(
      <CartStateProvider>
        <MockedProvider mocks={noSignedInMocks}>
          <Nav />
        </MockedProvider>
      </CartStateProvider>
    );

    expect(container).toHaveTextContent(/sign in/i);
    expect(container).toMatchSnapshot();

    const link = screen.getByText(/sign in/i);

    expect(link).toHaveAttribute("href", "/signin")
  });

  it("Renders a full nav when signed in", async () => {
    const { container } = render(
      <CartStateProvider>
        <MockedProvider mocks={signedInMocks}>
          <Nav />
        </MockedProvider>
      </CartStateProvider>
    );

    await screen.findByText(/account/i);

    expect(container).toHaveTextContent(/sign out/i)
    expect(container).toMatchSnapshot();
  })

  it("Renders the amount of items in the cart", async () => {
    const { container } = render(
      <CartStateProvider>
        <MockedProvider mocks={signedInMocksWithCartItems}>
          <Nav />
        </MockedProvider>
      </CartStateProvider>
    );

    await screen.findByText(/account/i);

    expect(screen.getByText("3")).toBeInTheDocument();
  })
});
