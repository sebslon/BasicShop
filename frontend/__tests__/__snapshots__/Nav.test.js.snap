// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Nav /> Renders a full nav when signed in 1`] = `
<div>
  <ul
    class="StyledNav-dh7aev-0 jJjMNl"
  >
    <a
      href="/products"
    >
      Products
    </a>
    <a
      href="/sell"
    >
      Sell
    </a>
    <a
      href="/orders"
    >
      Orders
    </a>
    <a
      href="/account"
    >
      Account
    </a>
    <button
      type="button"
    >
      My Cart
      <span
        class="CartCount__AnimationStyles-sc-1thwzyj-1 XjpSt"
      >
        <div>
          <div
            class="CartCount__Dot-sc-1thwzyj-0 bTkECI count"
          >
            0
          </div>
        </div>
      </span>
    </button>
    <button
      type="button"
    >
      Sign Out
    </button>
  </ul>
</div>
`;

exports[`<Nav /> Renders minimal nav and sign in link when signed out 1`] = `
<div>
  <ul
    class="StyledNav-dh7aev-0 jJjMNl"
  >
    <a
      href="/products"
    >
      Products
    </a>
    <a
      href="/signin"
    >
      Sign In
    </a>
  </ul>
</div>
`;
